RecoParticleFlow/PFClusterProducer/plugins/SpikeAndDoubleSpikeCleaner.cc

```diff
@@ -117,10 +117,65 @@ clean(const edm::Handle<reco::PFRecHitCollection>& input,
     }    
     const spike_cleaning& clean = _thresholds.find(hitlayer)->second;    
     if( rechit.energy() < clean._singleSpikeThresh ) continue;
+
+
+
+    //Fix needed for HF.  Here, we find the (up to) five companion rechits 
+    //to work in conjunction with the neighbours4() implementation below for the full HF surrounding energy
+    float compsumE = 0.0;
+    if ((hitlayer==PFLayer::HF_EM || hitlayer==PFLayer::HF_HAD))
+	{
+
+	int comp = 1;
+    	if (hitlayer==PFLayer::HF_EM) comp = 2;
+
+	int predphi =2;
+
+	//int comp1 = std::abs(hitlayer-13);
+
+	const HcalDetId& detid = (HcalDetId)rechit.detId();
+	std::vector<uint32_t> rawDetIds;
+	int heta = detid.ieta();
+	int hphi = detid.iphi();
+	
+	//At eta>39, phi segmentation changes 
+	if (std::abs(heta)>39) predphi=4;
+
+ 	int curphiL = hphi-predphi;
+	int curphiH = hphi+predphi;
+
+	//HcalDetId valid phi range (1-72)
+	if (predphi>abs(hphi) or predphi>abs(72-hphi))
+		{
+		while (curphiL-predphi<0) curphiL+=72;
+		while (curphiH+predphi>72) curphiH-=72;
+		}
+
+	std::pair<std::vector<int>, std::vector<int>>  phietas({heta,heta+1,heta-1,heta,heta},{hphi,hphi,hphi,curphiL,curphiH});
+	for(int in=0;in<5;in++)	
+		{
+        	HcalDetId tempID (HcalForward, phietas.first[in], phietas.second[in], comp);
+		rawDetIds.push_back(tempID.rawId());
+		}
+
+	for( const auto& jdx : ordered_hits ) 
+		{
+			const unsigned j = jdx;
+			const reco::PFRecHit& matchrechit = hits[j];
+  			for( const auto& iID : rawDetIds ) if (iID==matchrechit.detId())compsumE+=matchrechit.energy();  
+
+		}
+
+	}
+
+
+
+
+
     const double rhenergy = rechit.energy();
     // single spike cleaning
     auto const & neighbours4 = rechit.neighbours4();
-    double surroundingEnergy = rechit.energy();
+    double surroundingEnergy = compsumE;//rechit.energy();
```

Comment:
Perhaps @cbernet can check and answer? Or can you contact any other "ECAL expert"? This point must be clarified before we can proceed.

In the part of the code that you did not touch there are the two local variables:
- "neighbourEnergy", which after your upadates is just the same as surroundingEnergy everywhere else than for the HF
- "layerEnergy", which is identically the same as "neighbourEnergy"
which are never used. Could you please check whether they were supposed to be used, and if not remove them from the code?